flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RefCountingFSDataInputStream.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.runtime.state.AbstractChannelStateHandle;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.SupplierWithException;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+@NotThreadSafe
+class RefCountingFSDataInputStream extends FSDataInputStream {
+
+	@NotThreadSafe
+	static class RefCountingFSDataInputStreamFactory {
+		private final Map<StreamStateHandle, RefCountingFSDataInputStream> map = new HashMap<>(); // not clearing: expecting short life
+		private final ChannelStateDeserializer serializer;
+
+		RefCountingFSDataInputStreamFactory(ChannelStateDeserializer serializer) {
+			this.serializer = serializer;
+		}
+
+		<T> RefCountingFSDataInputStream forHandle(AbstractChannelStateHandle<T> handle) {
+			StreamStateHandle streamStateHandle = handle.getDelegate();
+			RefCountingFSDataInputStream stream = map.get(streamStateHandle);
+			if (stream == null) {
+				stream = new RefCountingFSDataInputStream(streamStateHandle::openInputStream, serializer);
+				map.put(streamStateHandle, stream);
+			}
+			return stream;
+		}
+
+		ChannelStateDeserializer getSerializer() {
+			return serializer;
+		}
+	}
+
+	private enum State {NEW, OPENED, CLOSED}
+
+	private final SupplierWithException<FSDataInputStream, IOException> streamSupplier;
+	private FSDataInputStream stream;
+	private final ChannelStateDeserializer serializer;
+	private int numReaders = 0;
+	private State state = State.NEW;
+
+	private RefCountingFSDataInputStream(SupplierWithException<FSDataInputStream, IOException> streamSupplier, ChannelStateDeserializer serializer) {
+		this.streamSupplier = streamSupplier;
+		this.serializer = serializer;
+	}
+
+	void incNumReaders() {
```

Comment:
nit: `incRef` seems more fit into the class name. also `numReaders` -> `refCounter`